frame/evm/precompile/dispatch/src/lib.rs

```diff
@@ -81,14 +81,30 @@ where
 			return Err(err);
 		}
 
+		#[cfg(feature = "evm-with-weight-limit")]
+		handle
+			.record_external_cost(Some(info.weight.ref_time()), Some(info.weight.proof_size()))?;
+
 		match call.dispatch(Some(origin).into()) {
 			Ok(post_info) => {
 				if post_info.pays_fee(&info) == Pays::Yes {
-					let cost = T::GasWeightMapping::weight_to_gas(
-						post_info.actual_weight.unwrap_or(info.weight),
-					);
-
+					let actual_weight = post_info.actual_weight.unwrap_or(info.weight);
+					let cost = T::GasWeightMapping::weight_to_gas(actual_weight);
 					handle.record_cost(cost)?;
+
+					#[cfg(feature = "evm-with-weight-limit")]
```

Comment:
I think we don't need to do like this, `record_external_cost` first, then `refund_external_cost` after call dispatched.  

We can deal with the cost in one place `let cost = T::GasWeightMapping::weight_to_gas(actual_weight);`,  Update the current `weight_to_gas` to take into account the `proof_size` field, then you do not need to refund any more. Futhermore, we don't need to record the external cost  before dispatching, just expand the `valid_weight` check to include the proof_size is enough IMO.

```rs
fn weight_to_gas(weight: Weight) -> u64 {
    weight.div(T::WeightPerGas::get().ref_time()).ref_time()
}
```

Reply:
@boundless-forest no we can't, the only way to properly accound for a resource usage is to refund after, otherwise we will account too many in some cases, which does not allow to manage the blockspace efficiently.